File: binary.js
class BinaryTree {
  constructor(data) {
    this.data = data;
    this.left = null;
    this.right = null;
  }
  addLeft(data) {
    this.left = data instanceof BinaryTree ? data : new BinaryTree(data);
    return this;
  }
  addRight(data) {
    this.right = data instanceof BinaryTree ? data : new BinaryTree(data);
    return this;
  }
  postOrder() {
    let output = this.data;
    let left = this.left == null ? "" : this.left.postOrder();
    let right = this.right == null ? "" : this.right.postOrder();
    return left + right + output;
  }
  inOrder() {
    let output = this.data;
    let left = this.left == null ? "" : this.left.inOrder();
    let right = this.right == null ? "" : this.right.inOrder();
    return left + output + right;
  }
  preOrder() {
    let output = this.data;
    let left = this.left == null ? "" : this.left.preOrder();
    let right = this.right == null ? "" : this.right.preOrder();
    return output + left + right;
  }
  breadthFirst() {
    return this.bfs(this);
  }
  bfs(node, queue = [node]) {
    if (node === undefined) return "";
    let shifted = queue.shift();
    let output = shifted.data;
    if (node.left !== null) queue.push(node.left);
    if (node.right !== null) queue.push(node.right);
    return output + this.bfs(queue[0], queue);
  }
  height() {
    let leftHeight = this.left === null ? 0 : this.left.height(),
      rightHeight = this.right === null ? 0 : this.right.height();
    return 1 + Math.max(leftHeight, rightHeight);
  }
}

const values = "FDJBEGKACIH".split(""),
  tree = values.map(v => new BinaryTree(v)),
  [f, d, j, b, e, g, k, a, c, i, h] = tree;

f.addLeft(d).addRight(j);
d.addLeft(b).addRight(e);
b.addLeft(a).addRight(c);
j.addLeft(g).addRight(k);
g.addRight(i);
i.addLeft(h);

const tree1 = f;

console.log("Breadth-First: ", f.breadthFirst());
console.log("Pre-Order: ", f.preOrder());
console.log("In-Order: ", f.inOrder());
console.log("Post-Order: ", f.postOrder());

File: binary.rb
# Ruby implementation of Binary Trees
class BinaryTree

  attr_reader :left, :right, :data, :height
  def initialize(data)
    @data = data
    @left = nil
    @right = nil
  end

  def add_left(data)
    @left = data.class == BinaryTree ? data : BinaryTree.new(data)
    self
  end

  def add_right(data)
    @right = data.class == BinaryTree ? data : BinaryTree.new(data)
    self
  end

  def post_order
    output = @data.nil? ? "" : @data
    left = @left.nil? ? "" : @left.post_order

    right = @right.nil? ? "" : @right.post_order

    left + right + output
  end

  def in_order
    output = @data.nil? ? "" : @data
    left = @left.nil? ? "" : @left.in_order
    right = @right.nil? ? "" : @right.in_order
    left + output + right
  end

  def pre_order
    output = @data.nil? ? "" : @data
    left = @left.nil? ? "" : @left.pre_order
    right = @right.nil? ? "" : @right.pre_order
    output + left + right
  end

  def breadth_first
    self.bfs(self)
  end

  def bfs(node, queue = [node])
    return "" if node.nil?
    shifted = queue.shift
    output = shifted.data.nil? ? "" : shifted.data
    queue.push(node.left) if !node.left.nil?
    queue.push(node.right) if !node.right.nil?
    output + self.bfs(queue[0], queue)
  end

  def height
    left_height = @left.nil? ? 0 : @left.height
    right_height = @left.nil? ? 0 : @right.height
    1 + [left_height,right_height].max
  end

end

values = "FDJBEGKACIH".split("")
tree = values.map {|value| BinaryTree.new(value)}

f, d, j, b, e, g, k, a, c, i, h = tree

f.add_left(d).add_right(j);
d.add_left(b).add_right(e);
b.add_left(a).add_right(c);
j.add_left(g).add_right(k);
g.add_right(i);
i.add_left(h);

tree1 = f

puts "Breadth-first: #{tree1.breadth_first}"
puts "Pre-Order: #{tree1.pre_order}"
puts "In-Order: #{tree1.in_order}"
puts "Post-OrderL #{tree1.post_order}"
puts tree1.height
